File: goodstein.rb
# 2014 - Ahmet Cetinkaya

def increase_base(number, base)
  new_base = base + 1
  new_number = 0

  i = 0
  while number > 0
    new_number += (number % base) * (new_base ** increase_base(i, base))
    number /= base
    i += 1
  end

  new_number
end


def iterate(number, base)
  increase_base(number, base) - 1
end


def goodstein_sequence(initial_number, initial_base = 2)
  sequence = [initial_number]

  number = initial_number
  base = initial_base
  while true
    number = iterate(number, base)
    sequence<< number
    base += 1

    break if number == 0
  end

  sequence
end

File: goodstein.rs
// 2015 - Ahmet Cetinkaya

use std::num::pow;

fn increase_base_aux(number: uint, base: uint, i: uint, acc: uint) -> uint {
    if number == 0 {
        acc
    } else {
        increase_base_aux(number / base, base, i + 1, acc + (number % base) * pow(base+1, increase_base(i, base)))
    }
}

fn increase_base(number: uint, base: uint) -> uint {
    increase_base_aux(number, base, 0, 0)
}

fn iterate(number: uint, base: uint) -> uint {
    increase_base(number, base) - 1
}

fn goodstein_sequence(initial_number: uint, initial_base: uint) -> Vec<uint> {
    let mut seq: Vec<uint> = vec![];
    let mut number = initial_number;
    let mut base = initial_base;
    while number > 0 {
        seq.push(number);
        number = iterate(number, base);
        base +=1;
    }
    seq.push(number);
    seq
}

fn main() {
    // expecting to see 6 6 6 6 5 4 3 2 1 0
   
    for i in goodstein_sequence(6, 4).iter() {
        println!("{}", i);
    }
}
